lib/msf/core/exploit/remote/ldap/server.rb

```diff
@@ -0,0 +1,127 @@
+
+module Msf
+
+###
+#
+# This module exposes methods for querying a remote LDAP service
+#
+###
+module Exploit::Remote::LDAP
+module Server
+  include Exploit::Remote::SocketServer
+
+  #
+  # Initializes an exploit module that serves LDAP requests
+  #
+  def initialize(info = {})
+    super
+
+    register_options(
+      [
+        OptPort.new('SRVPORT', [true, 'The local port to listen on.', 389]),
+        OptPath.new('LDIF_FILE', [ false, "Directory LDIF file path"]),
+      ], Exploit::Remote::LDAP::Server
+    )
+
+    register_advanced_options(
+      [
+        OptBool.new('LdapServerUdp', [true, "Serve UDP LDAP requests", true]),
+        OptBool.new('LdapServerTcp', [true, "Serve TCP LDAP requests", true])
+      ], Exploit::Remote::LDAP::Server
+    )
+  end
+
+  attr_accessor :service # :nodoc:
+
+  #
+  # Read LDIF file - from net-ldap ldapserver.rb#L162
+  #
+  # @ return [Hash] parsed ldif file
+  def read_ldif
+    if File.exists?(datastore['LDIF_FILE'])
+      ary = File.readlines(datastore['LDIF_FILE'])
+      ldif = {}
+      while (line = ary.shift) && line.chomp!
+        if line =~ /^dn:[\s]*/i
+          dn = $'
+          ldif[dn] = {}
+          while (attrib = ary.shift) && attrib.chomp! && attrib =~ /^([\w]+)[\s]*:[\s]*/
+            ldif[dn][$1.downcase] ||= []
+            ldif[dn][$1.downcase] << $'
+          end
+        end
+      end
+      return ldif
+    else
+      nil
+    end
+  end
+
+  #
+  # Handle incoming requests
+  # Override this method in modules to take flow control
+  #
+  def on_dispatch_request(cli, data)
+    service.default_dispatch_request(cli,data)
+  end
+
+  #
+  # Handle incoming requests
+  # Override this method in modules to take flow control
+  #
+  def on_send_response(cli, data)
+    cli.write(data)
+  end
+
+  #
+  # Starts the server
+  #
+  def start_service
+    begin
+      comm = _determine_server_comm(datastore['SRVHOST'])
+      self.service = Rex::ServiceManager.start(
+        Rex::Proto::LDAP::Server,
+        datastore['SRVHOST'],
+        datastore['SRVPORT'],
+        datastore['LdapServerUdp'],
+        datastore['LdapServerTcp'],
+        read_ldif,
+        comm,
+        {'Msf' => framework, 'MsfExploit' => self}
+      )
+
+      self.service.dispatch_request_proc = Proc.new do |cli, data|
+        on_dispatch_request(cli,data)
+      end
+      self.service.send_response_proc = Proc.new do |cli, data|
+        on_send_response(cli,data)
+      end
+      self.service.start
+
+    rescue ::Errno::EACCES => e
+      raise Rex::BindFailed.new(e.message)
+    end
+  end
+
+  #
+  # Stops the server
+  # @param destroy [TrueClass,FalseClass] Dereference the server object
+  def stop_service(destroy = false)
+    Rex::ServiceManager.stop_service(self.service) if self.service
+    if destroy
+      self.service = nil if self.service
+    end
+  end
+
+  #
+  # Resets the LDAP server
+  #
+  def reset_service
+    stop_service(true)
+    start_service
+  end
+
```

Comment:
Not sure if this is the best approach, but when this mixin is used with a scanner and the module is duplicated with `#replicant` the service for some reason no long is a kind of `Rex::Service` which causes a crash when the module's `#cleanup` is called which stops the service.

I was able to get around that using:
```
  def replicant
    obj = super
    if obj.service
      obj.service.extend(Rex::Service)
      obj.service.ref
    end
    obj
  end
```

Reply:
That **is** a problem.
Not sure we want to hack around that in such a manner - framework commit logs are littered with the sins of hackers doing such things to "just make it work"

Reply:
Potential solution to opt out of the replicant pattern https://github.com/rapid7/metasploit-framework/pull/15971

Reply:
Marking this as resolved since we've escalated addressing it out of this PR.